service/operator/operatorkit.go

```diff
@@ -0,0 +1,146 @@
+package operator
+
+import (
+	microerror "github.com/giantswarm/microkit/error"
+)
+
+// Operator implements the building blocks of any operator business logic being
+// reconciled when observing TPRs. This interface provides a guideline for a
+// more easy way to follow the rather complex intentions of operators in
+// general.
+type Operator interface {
+	// GetCurrentState receives the custom object observed during TPR watches. Its
+	// purpose is to return the current state of the resources being managed by
+	// the operator. This can e.g. be some actual data within a configmap as
+	// provided by the Kubernetes API. This is not limited to Kubernetes resources
+	// though. Another example would be to fetch and return information about
+	// Flannel bridges.
+	GetCurrentState(obj interface{}) (interface{}, error)
+	// GetDesiredState receives the custom object observed during TPR watches. Its
+	// purpose is to return the desired state of the resources being managed by
+	// the operator. The desired state should always be able to be made up using
+	// the information provided by the TPO. This can e.g. be some data within a
+	// configmap, how it should be provided by the Kubernetes API. This is not
+	// limited to Kubernetes resources though. Another example would be to make up
+	// and return information about Flannel bridges, how they should be look like
+	// on a server host.
+	GetDesiredState(obj interface{}) (interface{}, error)
+	// GetEmptyState is only to return the specific zero value the operator
+	// expects when reconciling delete operations. So this returns the desired
+	// state for delete operations. On create and delete events the operator might
+	// need to reconcile resources by removing them. GetDeleteState will receive
+	// as desiredState. This is to align to the general concept of reconciliation
+	// regardless creation or deletion events.
+	GetEmptyState() interface{}
+	// GetCreateState receives the custom object observed during TPR watches. It
+	// also receives the current state as provided by GetCurrentState and the
+	// desired state as provided by GetDesiredState. GetCreateState analyses the
+	// current and desired state and returns the state intended to be created by
+	// ProcessCreateState.
+	GetCreateState(obj, currentState, desiredState interface{}) (interface{}, error)
```

Comment:
I don't understand the naming. What is a `CreateState`? `CreateState` sounds more like an action than something I can `Get`/`Process` IMO.